storage/grpc_client.go

```diff
@@ -1721,21 +1728,323 @@ const (
+	if remainingInBuf < binary.MaxVarintLen64 {
+		log.Print("copying buf")
+		tagBuf = append(b[d.currOff:], ((*d.databufs)[d.currBuf+1].ReadOnlyData()[:binary.MaxVarintLen64-remainingInBuf])...)
+	}
+	return tagBuf
+}
+
+// This copies object data from the message into the buffer and returns the number of
+// bytes copied. The data offsets are incremented in the message. The updateCRC
+// function is called on the copied bytes.
+func (d *readResponseDecoder) readAndUpdateCRC(p []byte, updateCRC func([]byte)) int {
+	databuf := (*d.databufs)[d.dataOffsets.currBuf]
+	b := databuf.ReadOnlyData()
+	off := d.dataOffsets.currOff
+	n := copy(p, b[off:])
+	updateCRC(b[off:(off + uint64(n))])
+	d.dataOffsets.currOff += uint64(n)
+	// We are at the end of the current buffer
+	if d.dataOffsets.currOff == uint64(databuf.Len()) {
+		d.dataOffsets.currOff = 0
+		// We've read all the data from this message. Free the underlying buffers.
+		if d.dataOffsets.currBuf == d.dataOffsets.endBuf {
+			d.done = true
+			d.databufs.Free()
+		} else {
+			d.dataOffsets.currBuf++
+		}
+	}
+	return n
+}
+
+func (d *readResponseDecoder) writeToAndUpdateCRC(w io.Writer, updateCRC func([]byte)) (int64, error) {
+	var written int64
+	for !d.done {
+		databuf := (*d.databufs)[d.dataOffsets.currBuf]
+		b := databuf.ReadOnlyData()
+		off := d.dataOffsets.currOff
+		// Write all remaining data from the current buffer.
+		n, err := w.Write(b[off:])
+		written += int64(n)
+		updateCRC(b[off:])
+		if err != nil {
+			return written, err
+		}
+		d.dataOffsets.currOff = 0
+		// We've read all the data from this message.
+		if d.dataOffsets.currBuf == d.dataOffsets.endBuf {
+			d.done = true
+			d.databufs.Free()
+		} else {
+			d.dataOffsets.currBuf++
+		}
+	}
+	return written, nil
+}
+
+// Consume the next available tag in the input data and return the field number and type.
+// Advances the relevant offsets in the data.
+func (d *readResponseDecoder) consumeTag() (protowire.Number, protowire.Type, error) {
+	b := (*d.databufs)[d.currBuf].ReadOnlyData()
+	tagBuf := d.peek()
+
+	// Consume the next tag. This will tell us which field is next in the
+	// buffer, its type, and how much space it takes up.
+	fieldNum, fieldType, tagLength := protowire.ConsumeTag(tagBuf)
+	if tagLength < 0 {
+		return 0, 0, protowire.ParseError(tagLength)
+	}
+
+	remainingInBuf := len(b[d.currOff:])
+	// Update the offsets and current buffer depending on the tag length.
+	if tagLength >= remainingInBuf {
+		d.currBuf += 1
+		d.currOff = uint64(remainingInBuf) - uint64(tagLength)
+	} else {
+		d.currOff += uint64(tagLength)
+	}
+
+	d.off += uint64(tagLength)
+	return fieldNum, fieldType, nil
+}
+
+// Consume a varint that represents the length of a bytes field. Return the length of
+// the data, and advance the offsets by the length of the varint.
+func (d *readResponseDecoder) consumeVarint() (uint64, error) {
+	b := (*d.databufs)[d.currBuf].ReadOnlyData()
```

Comment:
I wonder if it's be worth stashing the current buffer's []byte in readResponseDecoder, to avoid the need for repeated calls to ReadOnlyData.

Or perhaps even stash a [][]byte and materialize all the buffers once at the start of decoding. You could avoid an allocation in the common case by keeping a [N][]byte in the decoder:

```
type readResponseDecoder struct {
  bufs [][]byte
  staticBufs [4][]byte
}

func (d *readResponseDecoder) init() {
  d.bufs = d.staticBufs[:0]
  for _, b := range d.databufs {
    d.bufs = append(d.bufs, b.ReadOnlyData())
  }
}
```

Reply:
I don't fully understand this, why would staticBufs be of len 4?

I think I'll probably just leave this for now if that's okay, I don't see overhead from these calls.